File: project/Yope3D/src/audio/Source.java
package audio;

import org.joml.Vector3f;
import org.lwjgl.openal.AL10;
import org.lwjgl.openal.AL11;

//encapsulates the OpenAL definition of a source
//makes it easier to interact with sources
public class Source {
	// openal source id
	private int sourceID;
	// openal buffer id (id of the buffer of sound)
	private int bufferID;
	// world position & velocity of this source
	private Vector3f position;
	private Vector3f velocity;
	// sound object associated with this source
	private Sound sound;

	public Source(Sound sound, Vector3f position, Vector3f velocity) {
		this.position = position;
		this.velocity = velocity;
		this.sound = sound;

		// gen a dummy source
		sourceID = AL10.alGenSources();
		bufferID = AL10.alGenBuffers();

		// load audio data into a buffer
		AL10.alBufferData(bufferID, (sound.getChannels() == 2) ? (AL10.AL_FORMAT_STEREO16): (AL10.AL_FORMAT_MONO16),
				sound.getData(), sound.getSamplingRate());

		// associate the buffer we just put with the audio source
		AL10.alSourcei(sourceID, AL10.AL_BUFFER, bufferID);

		// set some source properties
		AL10.alSourcef(sourceID, AL10.AL_GAIN, 1.0f);
		AL10.alSourcef(sourceID, AL10.AL_PITCH, 1.0f);

		AL10.alSource3f(sourceID, AL10.AL_POSITION, position.x, position.y, position.z);
		AL10.alSource3f(sourceID, AL10.AL_VELOCITY, velocity.x, velocity.y, velocity.z);
		
		//set the distance model
		AL10.alDistanceModel(AL11.AL_INVERSE_DISTANCE_CLAMPED);
		float globalReferenceDistance = 1;
		float globalMaxDistance = Float.MAX_VALUE;
		
		AL10.alSourcef(sourceID, AL10.AL_REFERENCE_DISTANCE, globalReferenceDistance);
		AL10.alSourcef(sourceID, AL10.AL_MAX_DISTANCE, globalMaxDistance);
	}
	
	public void play() {
		AL10.alSourcePlay(sourceID);
	}
	
	public void stop() {
		AL10.alSourceStop(sourceID);
	}
	
	public void pause() {
		AL10.alSourcePause(sourceID);
	}
	
	public void enableLooping() {
		AL10.alSourcei(sourceID, AL11.AL_LOOPING, AL10.AL_TRUE);
	}
	
	public void offset(int sec) {
		AL10.alSourcef(sourceID, AL11.AL_SEC_OFFSET, sec);
	}
	
	public void setGain(float f) {
		AL10.alSourcef(sourceID, AL10.AL_GAIN, f);
	}
	
	public void setPitch(float f) {
		AL10.alSourcef(sourceID, AL10.AL_PITCH, f);
	}
	
	public float getPitch() {
		return AL10.alGetSourcef(sourceID, AL10.AL_PITCH);
	}
	
	public float getGain() {
		return AL10.alGetSourcef(sourceID, AL10.AL_GAIN);
	}
	
	public void setPosition(Vector3f position) {
		AL10.alSource3f(sourceID, AL10.AL_POSITION, position.x, position.y, position.z);
		this.position = position;
	}
	
	public void setVelocity(Vector3f velocity) {
		AL10.alSource3f(sourceID, AL10.AL_VELOCITY, velocity.x, velocity.y, velocity.z);
		this.velocity = velocity;
	}
	
	public Sound getSound() {
		return sound;
	}
	
	public Vector3f getPosition() {
		return position;
	}
	
	public Vector3f getVelocity() {
		return velocity;
	}
	
	public void rewind() {
		AL10.alSourceRewind(sourceID);
	}
	
	public boolean isPlaying() {
		return AL10.alGetSourcei(sourceID, AL10.AL_SOURCE_STATE) == AL10.AL_PLAYING; 
	}
	
	public void cleanup() {
		AL10.alDeleteSources(sourceID);
		AL10.alDeleteBuffers(bufferID);
	}
}

File: v2/Yope3D/src/audio/Source.cpp
#include "Source.h"

Source::Source(ALuint bufferId) {
    alGenSources(1, &id_);
    alSourcei(id_, AL_BUFFER,           static_cast<ALint>(bufferId));
    alSourcei(id_, AL_SOURCE_RELATIVE,  AL_FALSE);   // absolute world-space position
    alSourcef(id_, AL_REFERENCE_DISTANCE, 1.0f);     // full volume within 1 unit
    alSourcef(id_, AL_ROLLOFF_FACTOR,     1.0f);
    alSourcef(id_, AL_MAX_DISTANCE,       1e9f);
}

Source::~Source() {
    if (id_) {
        alSourceStop(id_);
        alDeleteSources(1, &id_);
    }
}

void Source::play()   { alSourcePlay(id_);   }
void Source::pause()  { alSourcePause(id_);  }
void Source::stop()   { alSourceStop(id_);   }
void Source::rewind() { alSourceRewind(id_); }

void Source::setGain(float gain)   { alSourcef(id_, AL_GAIN,  gain);  }
void Source::setPitch(float pitch) { alSourcef(id_, AL_PITCH, pitch); }

void Source::setPosition(math::Vec3 pos) {
    alSource3f(id_, AL_POSITION, pos.x, pos.y, pos.z);
}
void Source::setVelocity(math::Vec3 vel) {
    alSource3f(id_, AL_VELOCITY, vel.x, vel.y, vel.z);
}

void Source::setReferenceDistance(float dist) {
    alSourcef(id_, AL_REFERENCE_DISTANCE, dist);
}

void Source::enableLooping(bool loop) {
    alSourcei(id_, AL_LOOPING, loop ? AL_TRUE : AL_FALSE);
}

void Source::setBuffer(ALuint bufferId) {
    alSourceStop(id_);   // AL_BUFFER can only be set on a non-playing source
    alSourcei(id_, AL_BUFFER, static_cast<ALint>(bufferId));
}

bool Source::isPlaying() const {
    ALint state = AL_STOPPED;
    alGetSourcei(id_, AL_SOURCE_STATE, &state);
    return state == AL_PLAYING;
}
